feat(48/RLN-INTEREP-SPEC): initial draft (#564)
vacp2p/rfc: content/docs/rfcs/48/README.md

---
slug: 48
title: 48/RLN-INTEREP-SPEC
name: Interep as group management for RLN
status: raw
category:  
tags: rln
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
---

# Abstract

> This spec integrates [Interep](https://interep.link) into the [RLN](https://rfc.vac.dev/spec/32/) spec. 

Reviewer: This is for a separate issue but this should probably refer to RLNP2P (can fix-up later)

Reviewer: Also I just realized it is Interep, not Interrep. I always read it as Inter(operable)-rep(utation)

Author: Addressed in [76516c1](https://github.com/vacp2p/rfc/pull/564/commits/76516c179bb029bd584257e05a0c34572d756729)

> This spec integrates [Interep](https://interep.link) into the [RLN](https://rfc.vac.dev/spec/32/) spec. 

Reviewer: In abstract I'd mention motivation briefly, e.g. a one sentence summary of Motivation section (~ allow users to join groups w/o funds but bootstrapping with web2 identity)

Author: Addressed in [9de124a](https://github.com/vacp2p/rfc/pull/564/commits/9de124addb6e16f5ec6908336a9d26e29b056f7e)

---
slug: 48
title: 48/RLN-INTEREP-SPEC
name: Interep as group management for RLN
status: raw
category:  
tags: rln
editor: Aaryamann Challani <aaryamann@status.im>
contributors:
---

# Abstract

This spec integrates [Interep](https://interep.link) into the [RLN](https://rlnp2p.vac.dev/) spec. 
Interep is a group management protocol that allows for the creation of groups of users and the management of their membership. 
It is used to manage the membership of the RLN group.

> Interep ties in web2 identities for reputation, and sorts the users into groups based on their reputation.

Reviewer 2: ```suggestion
Interep ties in web2 identities with reputation, and sorts the users into groups based on their reputation score.
```

> For example, a Github user with over 100 followers is considered to be of "gold" reputation.

Reviewer 2: ```suggestion
For example, a GitHub user with over 100 followers is considered to have "gold" reputation.
```

> Interep uses [Semaphore](https://semaphore.appliedzkp.org/) under the hood which allows anonymous signaling of membership in a group.

Reviewer 2: ```suggestion
Interep uses [Semaphore](https://semaphore.appliedzkp.org/) under the hood to allow anonymous signaling of membership in a group.
```

Reviewer 2: It is a bit unclear if semaphore is used to signal in a group (its original scope) or if you signal your _membership_ in a group (i.e. the fact that you have a certain membership)

> Therefore, a user with a "gold" reputation can signal their membership without revealing their identity.

Reviewer 2: It is unclear to me the meaning of "signal your membership"

Author: "Signal your membership" means "to convey that the user is part of a membership set"

Reviewer 2: I think my confusion comes from the wording used in Semaphore/RLN where signaling is a synonym for messaging. There you don't actually "signal" your membership, you "signal" a signal message and attach a proof that ensures your membership which allows your message to be forwarded further.

Author: Addressed in [82aa2e9](https://github.com/vacp2p/rfc/pull/564/commits/82aa2e909b6e41b74fcc7b3eb274cbd9a34a517d)

> RLN is used for spam prevention, and Interep is used for group management.

Reviewer 2: Is really the case that Interep does group management? My understanding is that it just certifies a group manager (e.g. a smart contract) that the registering user has a certain reputation.

Author: Well, yes. The semaphore groups are managed by Interep. We merely validate it.

By using Interep with RLN, we allow users to join RLN membership groups without the need for on-chain financial stake.

# Motivation

To have Sybil-Resistant group management, there are [implementations](https://github.com/vacp2p/rln-contract) of RLN which make use of financial stake on-chain.
However, this is not ideal because it reduces the barrier of entry for honest participants.

> In this case, honest participants will most likely have a web2 reputation accessible to them, which can be used for membership into an Interep reputation group.

Reviewer 2: ```suggestion
In this case, honest participants will most likely have a web2 identity accessible to them, which can be used for joining an Interep reputation group.
```

This proof is generated by the user, and is verified by the contract.
The contract ensures that the user is a member of the Interep group, and then adds them to the RLN membership group.

Following is the modified signature of the register function in the RLN contract -

```solidity
    /// @param groupId: Id of the group.
    /// @param signal: Semaphore signal.
    /// @param nullifierHash: Nullifier hash.
    /// @param externalNullifier: External nullifier.
    /// @param proof: Zero-knowledge proof.
    /// @param idCommitment: ID Commitment of the member.
    function register(
        uint256 groupId,
        bytes32 signal,
        uint256 nullifierHash,
        uint256 externalNullifier,
        uint256[8] calldata proof,
        uint256 idCommitment
    )
```

# Verification of messages

Messages are verified the same way as in the [RLN spec](https://rfc.vac.dev/spec/32/#verification).

# Slashing

The slashing mechanism is the same as in the [RLN spec](https://rfc.vac.dev/spec/32/#slashing).
It is important to note that the slashing may not have the intended effect on the user, since the only consequence is that they cannot send messages.

> This is due to the fact that the user can send a identity commitment in the registration to the RLN contract, which is different than the one used in the Interep group.

Reviewer 2: The interep proof should be zk, hence -in principle- reveals nothing about the user identity commitment. If this is true, then the same interep membership proof can be used for multiple registrations to the RLN contract. Is that the case?

Author: Oh! after further review, that does seem to be the case. Semaphore recommends to add a check as referenced [here](https://github.com/semaphore-protocol/semaphore/commit/f57e21c77d24a55d8b2f5a49b2a38c58792c6981#diff-32da6cb1264fc74594af2ac5e8d3abb032a5c6a8ddb589944bf4e86bed616d66) to prevent double signaling. I'll add this to the security considerations section

Author: Added in [d37babe](https://github.com/vacp2p/rfc/pull/564/commits/d37babe22a543930fbd572fac99274ef62d2e8ca)
